src/generateResponseType.js

```diff
@@ -338,48 +357,59 @@ export const unionOrInterfaceToFlow = (
 ): BabelNodeFlowType => {
-    const selectedAttributes: Array<
-        Array<BabelNodeObjectTypeProperty | BabelNodeObjectTypeSpreadProperty>,
-    > = type.possibleTypes
+    const allFields = selections.every(
+        (selection) => selection.kind === 'Field',
+    );
+    const selectedAttributes: Array<{
+        attributes: Array<
+            BabelNodeObjectTypeProperty | BabelNodeObjectTypeSpreadProperty,
+        >,
+        typeName: string,
+    }> = type.possibleTypes
         .slice()
         .sort((a, b) => {
             return a.name < b.name ? -1 : 1;
         })
         .map((possible) => {
             let seenTypeName = false;
-            return selections
-                .map((selection) =>
-                    unionOrInterfaceSelection(
-                        config,
-                        type,
-                        possible,
-                        selection,
-                    ),
-                )
-                .flat()
-                .filter((type) => {
-                    // The apollo-utilities "addTypeName" utility will add it
-                    // even if it's already specified :( so we have to filter out
-                    // the extra one here.
-                    if (
-                        type.type === 'ObjectTypeProperty' &&
-                        type.key.name === '__typename'
-                    ) {
-                        if (seenTypeName) {
-                            return false;
+            return {
+                attributes: selections
+                    .map((selection) =>
+                        unionOrInterfaceSelection(
+                            {
+                                ...config,
+                                path: allFields
+                                    ? config.path
+                                    : config.path.concat([possible.name]),
+                            },
+                            type,
+                            possible,
+                            selection,
+                        ),
+                    )
+                    .flat()
+                    .filter((type) => {
+                        // The apollo-utilities "addTypeName" utility will add it
+                        // even if it's already specified :( so we have to filter out
+                        // the extra one here.
+                        if (
+                            type.type === 'ObjectTypeProperty' &&
+                            type.key.name === '__typename'
+                        ) {
+                            if (seenTypeName) {
+                                return false;
+                            }
+                            seenTypeName = true;
                         }
-                        seenTypeName = true;
-                    }
-                    return true;
-                });
+                        return true;
+                    }),
+                typeName: possible.name,
```

Comment:
Nit: I think this would be easier to read if `typeName` were first, since the definition is short. That would make it easier to see that `typeName` and `attributes` are properties of the same object.